agent/taskresource/envFiles/envfile.go

```diff
@@ -0,0 +1,198 @@
+package envFiles
+
+import (
+	"sync"
+	"time"
+
+	"path/filepath"
+	"strings"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	"github.com/aws/amazon-ecs-agent/agent/credentials"
+	"github.com/aws/amazon-ecs-agent/agent/s3"
+	"github.com/aws/amazon-ecs-agent/agent/s3/factory"
+	resourcestatus "github.com/aws/amazon-ecs-agent/agent/taskresource/status"
+	"github.com/aws/amazon-ecs-agent/agent/utils/oswrapper"
+	"github.com/cihub/seelog"
+	"github.com/pkg/errors"
+
+	"github.com/aws/amazon-ecs-agent/agent/api/task/status"
+	"github.com/aws/amazon-ecs-agent/agent/taskresource"
+	"github.com/aws/amazon-ecs-agent/agent/utils/ioutilwrapper"
+)
+
+const (
+	envFileDirPath    = "envfiles"
+	envTempFilePrefix = "tmp_env"
+
+	s3DownloadTimeout = 30 * time.Second
+)
+
+// represents envfile as a task resource
+// these env files are retrieved from s3
+type EnvironmentFileResource struct {
+	cluster     string
+	taskARN     string
+	createdAt   time.Time
+	region      string
+	resourceDir string // path to store env var files
+
+	// env file related attributes
+	envfileLocations []apicontainer.EnvironmentFile // list of env file urls
+
+	executionCredentialsID string
+	credentialsManager     credentials.Manager
+	s3ClientCreator        factory.S3ClientCreator
+	os                     oswrapper.OS
+	ioutil                 ioutilwrapper.IOUtil
+
+	// Fields for the common functionality of task resource. Access to these fields are protected by lock.
+	createdAtUnsafe     time.Time
+	desiredStatusUnsafe resourcestatus.ResourceStatus
+	knownStatusUnsafe   resourcestatus.ResourceStatus
+	appliedStatusUnsafe resourcestatus.ResourceStatus
+	statusToTransitions map[resourcestatus.ResourceStatus]func() error
+	terminalReason      string
+	terminalReasonOnce  sync.Once
+	lock                sync.RWMutex
+}
+
+func NewEnvironmentFileResource(cluster, taskARN, region, dataDirOnHost string, credentialsManager credentials.Manager,
```

Comment:
i think where the code needs to write to is under ECS_DATADIR (by default /data), rather than under ECS_HOST_DATA_DIR (by default /var/lib/ecs),so dataDirOnHost should probably name as dataDir instead, and you don't really need to know the host dir here

Reply:
`/var/lib/ecs` is what was decided in the design doc. lmk concerns around this

Reply:
agent container has a bind mount `/var/lib/ecs/data:/data`. if you want to write to `/var/lib/ecs/data` on the host, you need to write to `/data` in the code.

Reply:
sure i'll change the name here and make sure to pass in ECS_DATADIR when this method is called